src/ucp/tag/tag_match.c

```diff
@@ -25,15 +41,75 @@ int ucp_tag_unexp_is_empty(ucp_tag_match_t *tm)
 
 void ucp_tag_exp_remove(ucp_tag_match_t *tm, ucp_request_t *req)
 {
+    ucs_queue_head_t *queue = ucp_tag_exp_get_queue(tm, req);
     ucs_queue_iter_t iter;
     ucp_request_t *qreq;
 
-    ucs_queue_for_each_safe(qreq, iter, &tm->expected, recv.queue) {
+    ucs_queue_for_each_safe(qreq, iter, queue, recv.queue) {
         if (qreq == req) {
-            ucs_queue_del_iter(&tm->expected, iter);
+            ucs_queue_del_iter(queue, iter);
             return;
         }
     }
 
     ucs_bug("expected request not found");
 }
+
+static inline uint64_t ucp_tag_exp_req_seq(ucs_queue_iter_t iter)
+{
+    return (*iter == NULL) ? ULONG_MAX :
```

Comment:
NULL means we reached the end of the queue, so i don't think this is likely/unlikely case